### PR TITLE
fix previous.js and the same function in Controller.js

### PR DESCRIPTION
**Pull request by cobianic**
## Please describe the changes this PR makes and why it should be merged:

When a player is created, the object Queue already contains `player.queue.previous` and it is identical to `player.queue.current`. Related bugs:

* every time the button "Replay" is pressed, a copy of the current playing song is added to the queue. Here I played a single song and pressed "Replay" button 2 times:
Controller_single_song_before

* If a queue of few tracks plays and `player.queue.current !== player.queue.previous`,  when using `/previous`, current and perious tracks start rotating so it could be a situation, when after using command `/previous` playing order instead of __previous->current->next1->next2__  is __previous->next1->next2__

* If a queue of few tracks plays and `player.queue.current !== player.queue.previous`,  when pressing the button "Replay", a copy of the current playing song is still added to the queue. Original queue:
before_original_queue

After I pressed "Replay" button 3 times:
Before_queue_after_pressing

## After this commit
A single song, pressed "Replay" button:
зображення

/previous, queue with songs
after_previous

Also  the current implementation of deleting messages after timeout throws errors "This interaction failed", so I want to revert this change.

### CODE REVIEW

**Comment by Saud-97** [comment]
What error do you get for sending the message and deleting it? It runs fine for me. Also try to make each PR focus on one topic for easy tracking and reverting if needed.

**Comment by cobianic** [comment]
For example, after my fix I started playing a single song and pressed "Previous" button. It showed an error message, as intended, but the button was thinking for 5 seconds and after showed this error from Discord
зображення

**Comment by Saud-97** [comment]
> For example, after my fix I started playing a single song and pressed "Previous" button. It showed an error message, as intended, but the button was thinking for 5 seconds and after showed this error from Discord. Overall everything worked (including delete after 5 seconds), it just showed errors зображення

This is an issue that was caused by your changes. I am pretty sure you have done something wrong which is why it shows.

Furthermore did you actually try to find the issue that caused in the logs and found the error or just changed it for no reason?
Perhaps your code is still throwing Discord API errors but it won't show in the Interaction since its only showing in logs.

**Comment by cobianic** [comment]
I`ve just recreated this error on the original client by playing something, pressing "Stop", then pressing "Play", so I am pretty sure that I`ve done nothing wrong, my changes just made these errors more obvious.

**Comment by Saud-97** [comment]
Can you at least show the logs what is the error that is happening? I can't really seem to figure out what's causing it.

**Comment by Saud-97** [comment]
Oh I see, this was probably unnoticed for a while 

The fix is easy just change the return of Stop and Replay.
from `return;` to `      return interaction.deferUpdate();
`
No need to change anything else.

**Comment by cobianic** [comment]
Yup, good catch. "Play" is also affected, other buttons are fine.

**Comment by Saud-97** [comment]
> Yup, good catch. "Play" is also affected, other buttons are fine.

Oh that was a typo. I meant `Replay `and `PlayAndPause `actions.

**Comment by Saud-97** [comment]
There is still few bugs. I will reply when I get to the PC and suggest the changes.

**Comment by Saud-97** [comment]
I just tested it. Hitting previous will only work once and then it will show this error although there are actually previous songs in the queue.

```
There is no previous song in the queue.
```

**Comment by cobianic** [comment]
It is not an error, the current queue can remember only 1 song behind and when the button or the command "Previous" is used, `player.queue.previous` becomes the same as `player.queue.current`, this is how the queue is realised in erela.js. I have thought about remembering the queue locally but didn`t try to implement it. If something, these changes should not be in conflict with a local copy of the queue, if such will be realized.

To satisfy researching interest, [here is the realisation of Queue in erela.js](https://github.com/MenuDocs/erela.js/blob/master/src/structures/Queue.ts)

**Comment by Saud-97** [comment]
Oh I thought your PR actually fixes previous and makes it work.

So what you did is instead of looping previous just make it stop looping? I mean that's not really a fix why bother do that.

It shouldn't be too hard to implement. Since the queue is already able to played via next track, I suggest to try and see if its doable to make it also work with previous. Otherwise I don't seen why is this PR needed tbh.

**Comment by cobianic** [comment]
Yes, this PR actually fixes bugs in previous. Previous was working even before. Bugs, that This PR fixes, are listed at the top. As for implementation, it will rewrite most of the commands and should be done as a separate PR.

**Comment by Saud-97** [comment]
Sure it should be a separate PR, I was just highlighting the more fundamental issue. #937. Anyways the PR looks good now for merge. Although you might consider keeping the formatting without replacing since it was just formatted in the latest commits so what's the point in reverting that. Notice you changed tabbed indents to spaces

File: util/Controller.js [hunk 1]
```diff
@@ -5,137 +5,146 @@ const { MessageEmbed } = require("discord.js");
...
+        player.pause(true);
+      }
+      client.warn(`Player: ${ player.options.guild } | Successfully ${ player.paused? "paused" : "resumed" } the player`);
+
+      interaction.update({
+        components: [client.createController(player.options.guild, player)],
+      });
+    }
+
+    return interaction.deferUpdate();
```
Comment (Saud-97) [inline util/Controller.js:121]: This is causing bugs. You can't combine 2 of these in the same code blocks.
```
interaction.reply() 
interaction.update()
interaction.deferUpdate()
```

File: util/Controller.js [hunk 2]
```diff
@@ -5,137 +5,146 @@ const { MessageEmbed } = require("discord.js");
...
+        embeds: [
+          new MessageEmbed()
+            .setColor("RED")
+            .setDescription("There is no song playing right now."),
+        ],
+      });
+      setTimeout(() => {
+        msg.delete();
+      }, 5000);
+
```
Comment (Saud-97) [inline util/Controller.js:106]:
```suggestion
return interaction.deferUpdate();
```

File: util/Controller.js (hunk 1 above)
Comment (Saud-97) [inline util/Controller.js:121]:
```suggestion

```

File: util/Controller.js [hunk 3]
```diff
@@ -5,137 +5,146 @@ const { MessageEmbed } = require("discord.js");
...
+    } else {
+
+      if (player.paused) {
+        player.pause(false);
+      } else {
+        player.pause(true);
+      }
+      client.warn(`Player: ${ player.options.guild } | Successfully ${ player.paused? "paused" : "resumed" } the player`);
+
+      interaction.update({
```
Comment (Saud-97) [inline util/Controller.js:116]:
```suggestion
     return interaction.update({
```

**Comment by cobianic** [comment]
Every change here is good, I've committed them, as for indentation, I don't know, how it happened. Code with your indentation is the same as in my IDE, and indents consist of two spaces, so I can't commit changes from IDE. Copying files manually to GitHub automatically add excessive indents. I've corrected previous.js manually by Shift-Tab, Controller.js for some reason fixed itself.

>Notice you changed tabbed indents to spaces

I don't quite understand you, indents can be Space(or multiple spaces) or Tab(or multiple Tabs). Even after your reformatting IDE shows that all indents consist of spaces, I don't see Tabs nor in IDE neither on github (at least for files previous.js and Controller.js)

**Comment by Saud-97** [comment]
It's fine if you still couldn't apply formatting. If you wanna see how it should look after formatting check this.
https://github.com/SudhanPlayz/Discord-MusicBot/pull/946/files#diff-0cd6077e2a9b32718203a28a9a3c363aef0be6a1516e863b91fb0bc842133933

**Comment by cobianic** [comment]
Got it, GitHub-side bug. Then, if I understand correctly, in the future commit every tab will become two spaces again, https://github.com/SudhanPlayz/Discord-MusicBot/commit/59c312792d36250a3959c0e93b4464d1ad4a26ae is a valid commit and there are no more questions to the indentation of this pull request, right?

**Comment by Saud-97** [comment]
> Got it, GitHub-side bug. Then, if I understand correctly, in the future commit every tab will become two spaces again, [59c3127](https://github.com/SudhanPlayz/Discord-MusicBot/commit/59c312792d36250a3959c0e93b4464d1ad4a26ae) is a valid commit and there are no more questions to the indentation of this pull request, right?

Well you just have to setup the indentation. I will create a PR to add support for a solution supported by IDEs to match formatting.

**Comment by Saud-97** [comment]
You can copy the same file in 35ed9e991ce75f4223e8ef35e72b33f72e38b650 and put it in the project folder. IDE should recognize the formatting configs from it.

**Comment by cobianic** [comment]
I`ve finally understood, that you have replaced all the spaces with a tab with 2 width. Well, I have converted and committed changes from two spaces to one tab with 2 width, but on Github it still looks like tabs with 4 width.

**Comment by Saud-97** [comment]
Ready to merge

**Comment by BioCla** [comment]
> @BioCla what this does again? I don't know all this, you should be the one who review this.

Honestly so much stuff has changed that it's almost completely deviated from what me, you and amiyo implemented.
I don't remember ever having had such bugs on the initial commit. But they say they fixed it, i guess we can trust them on it.
The code looks good afaik.

Its good to see people building upon the initial ideas either way, any further bugs can be reported and fixed at a later date anyway.

**Review by DarrenOfficial — APPROVED** [review]

**Pull request merged**